moxin-protocol/src/protocol.rs

```diff
@@ -1,21 +1,142 @@
+    pub use_mlock: bool,
+    pub n_batch: u32,
+    pub n_ctx: u32,
+    pub rope_freq_scale: f32,
+    pub rope_freq_base: f32,
 
-    // The argument is the File name.
-    DownloadFile(String),
+    // TBD Not really sure if this is something backend manages or if it is matter of
+    // the client (if it is done by tweaking the JSON payload for the chat completition)
+    pub context_overflow_policy: ContextOverflowPolicy
 }
 
 #[derive(Clone, Debug)]
-pub enum Response {
-    // Response to the GetFeaturedModels command
-    FeaturedModels(Vec<Model>),
+pub struct LoadedModelInfo {
+    pub file_id: FileID,
+    pub model_id: ModelID,
+
+    // JSON formatted string with the model information. See "Model Inspector" in LMStudio.
+    pub information: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct ModelResourcesInfo {
+    ram_usage: f32,
+    cpu_usage: f32,
+}
+
+#[derive(Clone, Debug)]
+pub enum LoadModelResponse {
+    Progress(FileID, f32),
+    Completed(LoadedModelInfo),
+    ModelResoucesUsage(ModelResourcesInfo),
+}
+
+pub enum ChatError {
+    BackendNotRun,
+    EndOfSequence,
+    ContextFull,
+    PromptTooLong,
+    TooLarge,
+    InvalidEncoding,
+    Other,
+}
+
+#[derive(Clone, Debug)]
+pub enum StopReason {
+    Completed,
+    Stopped
+}
+
+#[derive(Clone, Debug)]
+pub struct ChatCompletionData {
+    // The response from the model in JSON format
+    // https://platform.openai.com/docs/api-reference/chat/object
+    response: String,
+
+    // The remaining fields are stats about the chat completion process
+    time_to_first_token: f32,
+    time_to_generate: f32,
+    speed: f32,
+    gpu_layers: u32,
+    cpu_threads: u32,
+    mlock: bool,
+    token_count: u32,
+    token_limit: u32,
+    stop_reason: StopReason,
+}
+
+#[derive(Clone, Debug)]
+pub enum ChatResponse {
+    ChatCompletion(ChatCompletionData),
+    // https://platform.openai.com/docs/api-reference/chat/streaming
+    ChatCompletionChunk(String),
+}
+
+#[derive(Clone, Debug)]
+pub struct LocalServerConfig {
+    pub port: u16,
+    pub cors: bool,
+    pub request_queuing: bool,
+    pub verbose_server_logs: bool,
+    pub apply_prompt_formatting: bool,
+}
+
+#[derive(Clone, Debug)]
+pub enum LocalServerResponse {
+    Started,
+    Log(String),
+}
+
+#[derive(Clone, Debug)]
+pub enum Command {
+    GetFeaturedModels(Sender<Result<Vec<Model>>>),
+
+    // The argument is a string with the keywords to search for.
+    SearchModels(String, Sender<Result<Vec<Model>>>),
+
+    DownloadFile(FileID, Sender<Result<FileDownloadResponse>>),
+    GetDownloadedFiles(Sender<Result<Vec<DownloadedFile>>>),
+
+    LoadModel(FileID, LoadModelOptions, Sender<Result<LoadModelResponse>>),
+
+    // Eject currently loaded model, if any is provided
+    EjectModel(Sender<Result<()>>),
+
+    // The argument is the chat message in JSON format, following https://platform.openai.com/docs/api-reference/chat/create
+    Chat(String, Sender<Result<ChatResponse, ChatError>>),
```

Comment:
Same as ChatCompletionData::response. Could we use a struct that implements the serde to replace this string?

Reply:
That's a fair idea. Will change it 👍

Reply:
FYI I'm on hold with this change because I raised a requirements question on discord (visible to @juntao and others).

In LMStudio I have difference inference parameters than the ones defined in https://platform.openai.com/docs/api-reference/chat/create, so it's not clear to me which set we want to use.

In any case, let's change here to be explicit using rust structs.

Reply:
I decided to move forward with this idea, although I will confirm in tomorrow's call that our UI will use them.
Now OpenAI fields are part of structs/enums in our protocol.